File: be/internal/blueprint/blueprint.go
// Package blueprint defines the custom-test "blueprint": a declarative,
// versioned description of what a student asked for (filters, count, mix,
// mode, timing, marking). It is pure — no database access — so it can be
// validated, migrated and golden-tested in isolation.
package blueprint

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"strings"

	"codon-backend/internal/validate"

	"github.com/google/uuid"
)

const CurrentVersion = 1

type NCERT struct {
	Class    *int `json:"class,omitempty"`
	PageFrom *int `json:"page_from,omitempty"`
	PageTo   *int `json:"page_to,omitempty"`
}

type Filters struct {
	SubjectIDs                []uuid.UUID `json:"subject_ids,omitempty"`
	ChapterIDs                []uuid.UUID `json:"chapter_ids,omitempty"`
	TopicIDs                  []uuid.UUID `json:"topic_ids,omitempty"`
	Difficulty                []string    `json:"difficulty,omitempty"`
	Status                    []string    `json:"status,omitempty"`
	BookmarkCollectionIDs     []uuid.UUID `json:"bookmark_collection_ids,omitempty"`
	TagIDs                    []uuid.UUID `json:"tag_ids,omitempty"`
	ExcludeTagIDs             []uuid.UUID `json:"exclude_tag_ids,omitempty"`
	SourceTypes               []string    `json:"source_types,omitempty"`
	NCERT                     *NCERT      `json:"ncert,omitempty"`
	ExcludeAttemptedWithinDay int         `json:"exclude_attempted_within_days,omitempty"`
}

type Timing struct {
	Timed           bool `json:"timed"`
	DurationMinutes int  `json:"duration_minutes,omitempty"`
}

type Marking struct {
	Preset  string   `json:"preset,omitempty"`
	Correct *float64 `json:"correct,omitempty"`
	Wrong   *float64 `json:"wrong,omitempty"`
}

type Blueprint struct {
	SchemaVersion  int                `json:"schema_version"`
	CourseID       uuid.UUID          `json:"course_id"`
	Title          string             `json:"title,omitempty"`
	Filters        Filters            `json:"filters"`
	Count          int                `json:"count"`
	DifficultyMix  map[string]float64 `json:"difficulty_mix,omitempty"`
	SubjectWeights json.RawMessage    `json:"subject_weights,omitempty"` // "even" | "proportional" | {"<subject_id>": weight}
	Order          string             `json:"order,omitempty"`
	Strategy       string             `json:"strategy,omitempty"`
	Fallback       string             `json:"fallback,omitempty"` // "fewer" (default) | "fail"
	Mode           string             `json:"mode,omitempty"`
	Timing         Timing             `json:"timing"`
	Marking        Marking            `json:"marking"`
	Seed           *int64             `json:"seed,omitempty"`
}

var (
	Statuses   = []string{"unattempted", "incorrect", "correct", "bookmarked"}
	Orders     = []string{"random", "syllabus", "easy_first", "hard_first", "unattempted_first"}
	Strategies = []string{"random", "weak_first", "unseen_first", "spaced"}
	Fallbacks  = []string{"fewer", "fail"}
)

// Limits are the admin-tunable bounds a blueprint is validated against.
type Limits struct {
	MinQuestions, MaxQuestions int
	MinDuration, MaxDuration   int
	AllowedSources             []string // sources the pool may draw from
	TutorEnabled               bool
	StatusFiltersEnabled       bool
}

// Issue is one validation failure.
type Issue struct {
	Field   string `json:"field"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

func (i Issue) Error() string { return i.Field + ": " + i.Message }

// Parse decodes a blueprint. Unknown top-level fields are ignored (forward
// compatibility) but unknown keys inside `filters` are rejected. Old schema
// versions are migrated forward; a newer version than we know is an error.
func Parse(raw []byte) (Blueprint, []Issue) {
	var bp Blueprint
	if err := json.Unmarshal(raw, &bp); err != nil {
		return bp, []Issue{{"", "invalid_json", "blueprint is not valid JSON: " + err.Error()}}
	}
	var probe struct {
		Filters json.RawMessage `json:"filters"`
	}
	_ = json.Unmarshal(raw, &probe)
	if len(probe.Filters) > 0 && string(probe.Filters) != "null" {
		dec := json.NewDecoder(bytes.NewReader(probe.Filters))
		dec.DisallowUnknownFields()
		var f Filters
		if err := dec.Decode(&f); err != nil {
			return bp, []Issue{{"filters", "unknown_filter", "unsupported filter: " + err.Error()}}
		}
	}
	if bp.SchemaVersion == 0 {
		bp.SchemaVersion = 1
	}
	if bp.SchemaVersion > CurrentVersion {
		return bp, []Issue{{"schema_version", "unsupported_schema_version", fmt.Sprintf("blueprint version %d is newer than this server understands (%d)", bp.SchemaVersion, CurrentVersion)}}
	}
	bp.SchemaVersion = CurrentVersion
	return bp, nil
}

func maxIDs(field string, n int, issues *[]Issue) {
	if n > 200 {
		*issues = append(*issues, Issue{field, "too_many", "at most 200 ids allowed"})
	}
}

// Normalize validates the blueprint against lim and fills every default, so
// the stored/snapshotted blueprint is fully resolved.
func Normalize(bp *Blueprint, lim Limits) []Issue {
	var is []Issue
	add := func(field, code, msg string) { is = append(is, Issue{field, code, msg}) }

	if bp.CourseID == uuid.Nil {
		add("course_id", "required", "course_id is required")
	}
	if len([]rune(bp.Title)) > 120 {
		add("title", "too_long", "title must be at most 120 characters")
	}
	bp.Title = strings.TrimSpace(bp.Title)

	f := &bp.Filters
	maxIDs("filters.subject_ids", len(f.SubjectIDs), &is)
	maxIDs("filters.chapter_ids", len(f.ChapterIDs), &is)
	maxIDs("filters.topic_ids", len(f.TopicIDs), &is)
	maxIDs("filters.tag_ids", len(f.TagIDs), &is)
	maxIDs("filters.exclude_tag_ids", len(f.ExcludeTagIDs), &is)
	for _, d := range f.Difficulty {
		if !validate.OneOf(d, validate.Difficulties) {
			add("filters.difficulty", "bad_enum", "difficulty must be easy, medium or hard")
		}
	}
	for _, s := range f.Status {
		if !validate.OneOf(s, Statuses) {
			add("filters.status", "bad_enum", "status must be one of "+strings.Join(Statuses, ", "))
		}
	}
	if len(f.Status) > 0 && !lim.StatusFiltersEnabled {
		add("filters.status", "filter_unavailable", "question-status filters are not enabled yet")
	}
	if len(f.BookmarkCollectionIDs) > 0 && !validate.OneOf("bookmarked", f.Status) {
		add("filters.bookmark_collection_ids", "needs_bookmarked_status", "bookmark_collection_ids only applies together with status \"bookmarked\"")
	}
	// sources: intersect with what the pool is allowed to draw from
	for _, st := range f.SourceTypes {
		if !validate.OneOf(st, validate.SourceTypes) {
			add("filters.source_types", "bad_enum", "unknown source type: "+st)
		} else if !validate.OneOf(st, lim.AllowedSources) {
			add("filters.source_types", "source_not_allowed", st+" questions are not available for custom tests")
		}
	}
	if n := f.NCERT; n != nil {
		if n.PageFrom != nil && n.PageTo != nil && *n.PageFrom > *n.PageTo {
			add("filters.ncert", "bad_range", "page_from must be <= page_to")
		}
		if n.Class != nil && (*n.Class < 1 || *n.Class > 12) {
			add("filters.ncert.class", "out_of_range", "class must be between 1 and 12")
		}
	}
	if f.ExcludeAttemptedWithinDay < 0 || f.ExcludeAttemptedWithinDay > 365 {
		add("filters.exclude_attempted_within_days", "out_of_range", "must be between 0 and 365")
	}

	if bp.Count < lim.MinQuestions || bp.Count > lim.MaxQuestions {
		add("count", "out_of_range", fmt.Sprintf("count must be between %d and %d", lim.MinQuestions, lim.MaxQuestions))
	}
	if len(bp.DifficultyMix) > 0 {
		sum := 0.0
		for k, v := range bp.DifficultyMix {
			if !validate.OneOf(k, validate.Difficulties) || v < 0 || v > 1 {
				add("difficulty_mix", "invalid", "mix keys are easy/medium/hard with weights between 0 and 1")
			}
			sum += v
		}
		if math.Abs(sum-1) > 0.02 {
			add("difficulty_mix", "must_sum_to_one", "difficulty_mix weights must add up to 1")
		}
	}
	if len(bp.SubjectWeights) > 0 {
		var mode string
		if json.Unmarshal(bp.SubjectWeights, &mode) == nil {
			if mode != "even" && mode != "proportional" {
				add("subject_weights", "invalid", `subject_weights must be "even", "proportional" or a map of subject id → weight`)
			}
		} else {
			var m map[string]float64
			if err := json.Unmarshal(bp.SubjectWeights, &m); err != nil {
				add("subject_weights", "invalid", "subject_weights must be a map of subject id → weight")
			}
			for k, v := range m {
				if _, err := uuid.Parse(k); err != nil || v < 0 {
					add("subject_weights", "invalid", "weights are keyed by subject uuid with values >= 0")
				}
			}
		}
	}
	if bp.Order == "" {
		bp.Order = "random"
	}
	if !validate.OneOf(bp.Order, Orders) {
		add("order", "bad_enum", "order must be one of "+strings.Join(Orders, ", "))
	}
	if bp.Strategy == "" {
		bp.Strategy = "random"
	}
	if !validate.OneOf(bp.Strategy, Strategies) {
		add("strategy", "bad_enum", "strategy must be one of "+strings.Join(Strategies, ", "))
	}
	if bp.Fallback == "" {
		bp.Fallback = "fewer"
	}
	if !validate.OneOf(bp.Fallback, Fallbacks) {
		add("fallback", "bad_enum", "fallback must be fewer or fail")
	}
	if bp.Mode == "" {
		bp.Mode = "exam"
	}
	if !validate.OneOf(bp.Mode, validate.Modes) {
		add("mode", "bad_enum", "mode must be exam or tutor")
	} else if bp.Mode == "tutor" && !lim.TutorEnabled {
		add("mode", "mode_unavailable", "tutor mode is not enabled yet")
	}

	if bp.Timing.Timed {
		if bp.Timing.DurationMinutes < lim.MinDuration || bp.Timing.DurationMinutes > lim.MaxDuration {
			add("timing.duration_minutes", "out_of_range", fmt.Sprintf("duration must be between %d and %d minutes", lim.MinDuration, lim.MaxDuration))
		}
	} else {
		bp.Timing.DurationMinutes = 0
	}

	switch bp.Marking.Preset {
	case "", "neet":
		bp.Marking.Preset = "neet"
		c, w := 4.0, -1.0
		bp.Marking.Correct, bp.Marking.Wrong = &c, &w
	case "no_negative":
		c, w := 1.0, 0.0
		bp.Marking.Correct, bp.Marking.Wrong = &c, &w
	case "custom":
		if bp.Marking.Correct == nil || bp.Marking.Wrong == nil {
			add("marking", "required", "custom marking needs correct and wrong")
		} else if *bp.Marking.Correct <= 0 || *bp.Marking.Correct > 10 || *bp.Marking.Wrong > 0 || *bp.Marking.Wrong < -10 {
			add("marking", "out_of_range", "correct must be in (0, 10] and wrong in [-10, 0]")
		}
	default:
		add("marking.preset", "bad_enum", "preset must be neet, no_negative or custom")
	}
	return is
}

// Marks returns the resolved (correct, wrong) marks of a normalised blueprint.
func (b Blueprint) Marks() (float64, float64) {
	c, w := 4.0, -1.0
	if b.Marking.Correct != nil {
		c = *b.Marking.Correct
	}
	if b.Marking.Wrong != nil {
		w = *b.Marking.Wrong
	}
	return c, w
}

// SubjectWeightMode returns "even", "proportional", "custom" or "" (none).
func (b Blueprint) SubjectWeightMode() (string, map[uuid.UUID]float64) {
	if len(b.SubjectWeights) == 0 {
		return "", nil
	}
	var mode string
	if json.Unmarshal(b.SubjectWeights, &mode) == nil {
		return mode, nil
	}
	var m map[string]float64
	_ = json.Unmarshal(b.SubjectWeights, &m)
	out := map[uuid.UUID]float64{}
	for k, v := range m {
		if id, err := uuid.Parse(k); err == nil {
			out[id] = v
		}
	}
	return "custom", out
}

// JSON returns the canonical stored form.
func (b Blueprint) JSON() []byte {
	out, _ := json.Marshal(b)
	return out
}

File: fe/src/custom/blueprint.ts
import type { Blueprint, BlueprintFilters, BlueprintIssue, Difficulty } from '@/src/api/customTests';

/**
 * Blueprint state, defaults and validation for the custom-test builder.
 *
 * `validateBlueprint` mirrors the server's `blueprint.Normalize` and is tested
 * against the shared golden vectors (contracts/blueprint-v1/vectors.json), so
 * the builder can show field errors instantly and the two sides can't drift
 * apart unnoticed. The server stays the authority: it re-validates everything.
 */

export const CURRENT_SCHEMA_VERSION = 1;

export type Limits = {
  min_questions: number;
  max_questions: number;
  min_duration: number;
  max_duration: number;
  allowed_sources: string[];
  tutor_enabled: boolean;
  status_filters_enabled: boolean;
};

const DIFFICULTIES = ['easy', 'medium', 'hard'];
const STATUSES = ['unattempted', 'incorrect', 'correct', 'bookmarked'];
const ORDERS = ['random', 'syllabus', 'easy_first', 'hard_first', 'unattempted_first'];
const STRATEGIES = ['random', 'weak_first', 'unseen_first', 'spaced'];
const FILTER_KEYS = new Set([
  'subject_ids', 'chapter_ids', 'topic_ids', 'difficulty', 'status', 'bookmark_collection_ids', 'tag_ids',
  'exclude_tag_ids', 'source_types', 'ncert', 'exclude_attempted_within_days',
]);
const SOURCES = ['qbank', 'practice', 'pyq', 'test_series', 'other'];
const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export function defaultBlueprint(courseId: string, count = 20): Blueprint {
  return {
    schema_version: CURRENT_SCHEMA_VERSION,
    course_id: courseId,
    filters: {},
    count,
    mode: 'exam',
    timing: { timed: false },
    marking: { preset: 'neet' },
  };
}

/** Returns the issues the server would raise (same `field`/`code`). Empty = valid. */
export function validateBlueprint(input: any, lim: Limits): BlueprintIssue[] {
  const is: BlueprintIssue[] = [];
  const add = (field: string, code: string, message: string) => is.push({ field, code, message });
  const bp = input ?? {};

  if (bp.schema_version != null && bp.schema_version > CURRENT_SCHEMA_VERSION) {
    return [{ field: 'schema_version', code: 'unsupported_schema_version', message: `blueprint version ${bp.schema_version} is newer than this app understands` }];
  }
  if (bp.filters && typeof bp.filters === 'object') {
    for (const k of Object.keys(bp.filters)) {
      if (!FILTER_KEYS.has(k)) return [{ field: 'filters', code: 'unknown_filter', message: `unsupported filter: ${k}` }];
    }
  }

  if (!bp.course_id || bp.course_id === '00000000-0000-0000-0000-000000000000') add('course_id', 'required', 'course_id is required');
  if ([...(bp.title ?? '')].length > 120) add('title', 'too_long', 'title must be at most 120 characters');

  const f: BlueprintFilters = bp.filters ?? {};
  for (const [name, ids] of Object.entries({
    subject_ids: f.subject_ids, chapter_ids: f.chapter_ids, topic_ids: f.topic_ids, tag_ids: f.tag_ids, exclude_tag_ids: f.exclude_tag_ids,
  })) {
    if ((ids?.length ?? 0) > 200) add(`filters.${name}`, 'too_many', 'at most 200 ids allowed');
  }
  for (const d of f.difficulty ?? []) if (!DIFFICULTIES.includes(d)) add('filters.difficulty', 'bad_enum', 'difficulty must be easy, medium or hard');
  for (const s of f.status ?? []) if (!STATUSES.includes(s)) add('filters.status', 'bad_enum', `status must be one of ${STATUSES.join(', ')}`);
  if ((f.status?.length ?? 0) > 0 && !lim.status_filters_enabled) add('filters.status', 'filter_unavailable', 'question-status filters are not enabled yet');
  if ((f.bookmark_collection_ids?.length ?? 0) > 0 && !(f.status ?? []).includes('bookmarked')) {
    add('filters.bookmark_collection_ids', 'needs_bookmarked_status', 'bookmark_collection_ids only applies together with status "bookmarked"');
  }
  for (const st of f.source_types ?? []) {
    if (!SOURCES.includes(st)) add('filters.source_types', 'bad_enum', `unknown source type: ${st}`);
    else if (!lim.allowed_sources.includes(st)) add('filters.source_types', 'source_not_allowed', `${st} questions are not available for custom tests`);
  }
  const n = f.ncert;
  if (n) {
    if (n.page_from != null && n.page_to != null && n.page_from > n.page_to) add('filters.ncert', 'bad_range', 'page_from must be <= page_to');
    if (n.class != null && (n.class < 1 || n.class > 12)) add('filters.ncert.class', 'out_of_range', 'class must be between 1 and 12');
  }
  if (f.exclude_attempted_within_days != null && (f.exclude_attempted_within_days < 0 || f.exclude_attempted_within_days > 365)) {
    add('filters.exclude_attempted_within_days', 'out_of_range', 'must be between 0 and 365');
  }

  if (!(bp.count >= lim.min_questions && bp.count <= lim.max_questions)) {
    add('count', 'out_of_range', `count must be between ${lim.min_questions} and ${lim.max_questions}`);
  }

  const mix = bp.difficulty_mix as Record<string, number> | undefined;
  if (mix && Object.keys(mix).length > 0) {
    let sum = 0;
    let bad = false;
    for (const [k, v] of Object.entries(mix)) {
      if (!DIFFICULTIES.includes(k) || v < 0 || v > 1) bad = true;
      sum += v;
    }
    if (bad) add('difficulty_mix', 'invalid', 'mix keys are easy/medium/hard with weights between 0 and 1');
    if (Math.abs(sum - 1) > 0.02) add('difficulty_mix', 'must_sum_to_one', 'difficulty_mix weights must add up to 1');
  }

  const sw = bp.subject_weights;
  if (sw != null) {
    if (typeof sw === 'string') {
      if (sw !== 'even' && sw !== 'proportional') add('subject_weights', 'invalid', 'subject_weights must be "even", "proportional" or a map of subject id → weight');
    } else if (typeof sw === 'object' && !Array.isArray(sw)) {
      for (const [k, v] of Object.entries(sw as Record<string, number>)) {
        if (!UUID.test(k) || !(typeof v === 'number') || v < 0) {
          add('subject_weights', 'invalid', 'weights are keyed by subject uuid with values >= 0');
          break;
        }
      }
    } else add('subject_weights', 'invalid', 'subject_weights must be a map of subject id → weight');
  }

  if (bp.order && !ORDERS.includes(bp.order)) add('order', 'bad_enum', `order must be one of ${ORDERS.join(', ')}`);
  if (bp.strategy && !STRATEGIES.includes(bp.strategy)) add('strategy', 'bad_enum', `strategy must be one of ${STRATEGIES.join(', ')}`);
  if (bp.fallback && bp.fallback !== 'fewer' && bp.fallback !== 'fail') add('fallback', 'bad_enum', 'fallback must be fewer or fail');

  const mode = bp.mode || 'exam';
  if (mode !== 'exam' && mode !== 'tutor') add('mode', 'bad_enum', 'mode must be exam or tutor');
  else if (mode === 'tutor' && !lim.tutor_enabled) add('mode', 'mode_unavailable', 'tutor mode is not enabled yet');

  if (bp.timing?.timed) {
    const d = bp.timing.duration_minutes ?? 0;
    if (d < lim.min_duration || d > lim.max_duration) add('timing.duration_minutes', 'out_of_range', `duration must be between ${lim.min_duration} and ${lim.max_duration} minutes`);
  }

  const m = bp.marking ?? {};
  const preset = m.preset || 'neet';
  if (preset === 'custom') {
    if (m.correct == null || m.wrong == null) add('marking', 'required', 'custom marking needs correct and wrong');
    else if (m.correct <= 0 || m.correct > 10 || m.wrong > 0 || m.wrong < -10) add('marking', 'out_of_range', 'correct must be in (0, 10] and wrong in [-10, 0]');
  } else if (preset !== 'neet' && preset !== 'no_negative') add('marking.preset', 'bad_enum', 'preset must be neet, no_negative or custom');

  return is;
}

/** The (correct, wrong) marks a blueprint resolves to — mirrors `Blueprint.Marks()` after Normalize. */
export function resolvedMarks(bp: Blueprint): { correct: number; wrong: number } {
  switch (bp.marking?.preset ?? 'neet') {
    case 'no_negative': return { correct: 1, wrong: 0 };
    case 'custom': return { correct: bp.marking.correct ?? 4, wrong: bp.marking.wrong ?? -1 };
    default: return { correct: 4, wrong: -1 };
  }
}

/** Suggested duration: minutes-per-question × count, clamped to the allowed range. */
export function suggestedMinutes(count: number, perQuestion: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, Math.ceil(count * perQuestion)));
}

// ── immutable edit helpers (the builder reducer builds on these) ────────────

export function toggleInList<T>(list: T[] | undefined, value: T): T[] {
  const cur = list ?? [];
  return cur.includes(value) ? cur.filter((v) => v !== value) : [...cur, value];
}

/** Drop empty arrays/objects so equal filters serialise identically (stable count-cache keys, tidy blueprints). */
export function pruneFilters(f: BlueprintFilters): BlueprintFilters {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(f)) {
    if (Array.isArray(v) && v.length === 0) continue;
    if (v == null) continue;
    if (k === 'ncert' && typeof v === 'object' && Object.values(v).every((x) => x == null)) continue;
    if (k === 'exclude_attempted_within_days' && v === 0) continue;
    out[k] = v;
  }
  return out as BlueprintFilters;
}

/** Filter state as a string — the key for count-request dedupe/caching. */
export function blueprintKey(bp: Blueprint): string {
  return JSON.stringify({ ...bp, filters: pruneFilters(bp.filters ?? {}), title: undefined });
}

export function withDifficultyMix(bp: Blueprint, mix: Partial<Record<Difficulty, number>> | undefined): Blueprint {
  const next = { ...bp };
  if (!mix || Object.keys(mix).length === 0) delete next.difficulty_mix;
  else next.difficulty_mix = mix;
  return next;
}

/**
 * Chapter/subject/topic selection is hierarchical: picking a whole subject
 * makes its chapters redundant. `normalizeSelection` removes ids already
 * covered by a broader selection so the blueprint says one thing once.
 */
export function normalizeSelection(
  f: BlueprintFilters,
  tree: { id: string; chapters: { id: string; topics: { id: string }[] }[] }[],
): BlueprintFilters {
  const subj = new Set(f.subject_ids ?? []);
  const chap = new Set(f.chapter_ids ?? []);
  const chapterOfTopic = new Map<string, string>();
  const subjectOfChapter = new Map<string, string>();
  for (const s of tree) for (const c of s.chapters) {
    subjectOfChapter.set(c.id, s.id);
    for (const t of c.topics) chapterOfTopic.set(t.id, c.id);
  }
  const chapter_ids = (f.chapter_ids ?? []).filter((c) => !subj.has(subjectOfChapter.get(c) ?? ''));
  const chapKeep = new Set(chapter_ids);
  const topic_ids = (f.topic_ids ?? []).filter((t) => {
    const c = chapterOfTopic.get(t);
    return !(c && (chap.has(c) && chapKeep.has(c))) && !subj.has(subjectOfChapter.get(c ?? '') ?? '');
  });
  return pruneFilters({ ...f, chapter_ids, topic_ids });
}

/**
 * Apply a server suggestion/recommendation patch (a partial blueprint) to a
 * blueprint: top-level keys replace, `filters` merges key-by-key, and a `null`
 * value removes the key (e.g. `{"filters": {"ncert": null}}`).
 */
export function applyPatch(bp: Blueprint, patch: Record<string, unknown>): Blueprint {
  const next: Record<string, any> = { ...bp };
  for (const [k, v] of Object.entries(patch)) {
    if (k === 'filters' && v && typeof v === 'object') {
      const merged: Record<string, unknown> = { ...(bp.filters ?? {}) };
      for (const [fk, fv] of Object.entries(v as Record<string, unknown>)) {
        if (fv === null) delete merged[fk];
        else merged[fk] = fv;
      }
      next.filters = pruneFilters(merged as BlueprintFilters);
    } else if (v === null) delete next[k];
    else next[k] = v;
  }
  return next as Blueprint;
}

/** One-line human summary of a blueprint, for cards and share sheets. */
export function summarizeBlueprint(bp: Blueprint, names: { subjects?: Record<string, string> } = {}): string {
  const parts: string[] = [`${bp.count} Qs`];
  const subj = (bp.filters?.subject_ids ?? []).map((id) => names.subjects?.[id]).filter(Boolean) as string[];
  if (subj.length) parts.push(subj.slice(0, 2).join(', ') + (subj.length > 2 ? ` +${subj.length - 2}` : ''));
  else if ((bp.filters?.chapter_ids?.length ?? 0) > 0) parts.push(`${bp.filters.chapter_ids!.length} chapter${bp.filters.chapter_ids!.length > 1 ? 's' : ''}`);
  if (bp.filters?.difficulty?.length) parts.push(bp.filters.difficulty.join('/'));
  parts.push(bp.timing?.timed ? `${bp.timing.duration_minutes} min` : 'untimed');
  if (bp.mode === 'tutor') parts.push('tutor');
  return parts.join(' · ');
}
